In [7]:
import pandas as pd

Data Loading

In [1]:
import pandas as pd

file_path = 'online_retail.xlsx'  # Make sure this matches the name of your file

data = pd.read_excel(file_path)

print("data loaded")

data loaded


In [3]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

Data Processing

1. Handling Missing Values

In [4]:
print(data.isnull().sum())

data.dropna(inplace=True)

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64


The results above shows that there are null values in the Description and CustomerID columns. Description has 1454 rows with null values and CustomerID has 135080 values.

2. Remove Duplicates

In [5]:
data.drop_duplicates(inplace=True)

In [6]:
print(data.isnull().sum())

data.dropna(inplace=True)

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64


3. Data Transformation

Here we create three new columns, Year, Month, and Day to help with the analysis. It will amke it easier to uncover insights and build effective models.

In [8]:
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

data['Year'] = data['InvoiceDate'].dt.year
data['Month'] = data['InvoiceDate'].dt.month
data['Day'] = data['InvoiceDate'].dt.day

data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Year  Month  Day  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010     12    1  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12    1  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010     12    1  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12    1  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12    1

4. Handling Outliers

In [10]:
negative_quantities = data[data['Quantity'] < 0]
print("Number of negative quantities:", negative_quantities.shape[0])

non_positive_prices = data[data['UnitPrice'] <= 0]
print("Number of zero or negative unit prices:", non_positive_prices.shape[0])

Number of negative quantities: 8872
Number of zero or negative unit prices: 40


Negative quantities in this data mean errors in the data collection and data recording or products that were returned back to the company.

Negative unit prices in this data could mean products that were part of a giveaway or that were part of a discount period maybe sold at 0.00

4.1 Removing the negative quatities and the negative unit prices from the data set

In [12]:
data = data[data['Quantity'] > 0]

data = data[data['UnitPrice'] > 0]

data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Year  Month  Day  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010     12    1  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12    1  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010     12    1  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12    1  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010     12    1

In [13]:
negative_quantities = data[data['Quantity'] < 0]
print("Number of negative quantities:", negative_quantities.shape[0])

non_positive_prices = data[data['UnitPrice'] <= 0]
print("Number of zero or negative unit prices:", non_positive_prices.shape[0])

Number of negative quantities: 0
Number of zero or negative unit prices: 0
